File: src/builder.rb
require 'json'
require 'fileutils'
require 'time'

# Gems
require 'kramdown'
require 'fastimage'

class Builder

  OUTPUT_DIR = 'bin' # where to build to
  DATA_DIR = 'data' # where our data files are
  DATABASE_FILE = "#{DATA_DIR}/games.json"
  STATIC_PAGES_DIR = "#{DATA_DIR}/static_pages"
  IMAGES_DIR = "#{DATA_DIR}/images"
  GAMES_DIR = "#{DATA_DIR}/games"

  TEMPLATE_DIRECTORY = 'templates'
  # We copy the template dir. But not these items.
  TEMPLATE_EXCLUSIONS = ['snippets']

  INDEX_PAGE = 'index.html' # the home index page; this is also the layout.
  NAVBAR_LINKS_PLACEHOLDER = '<!-- DG navbar links -->' # Where the navbar list of pages goes
  # Snippets where we put in real data (eg. game page URLs)
  NAVBAR_LINK_SNIPPET = "#{TEMPLATE_DIRECTORY}/snippets/navbar_link.html"
  JUMBOTRON_SNIPPET = "#{TEMPLATE_DIRECTORY}/snippets/jumbotron.html"
  CONTENT_PLACEHOLDER = '<!-- DG content -->' # Where in the template we fill in the page content
  GAME_PAGE_TEMPLATE = "#{TEMPLATE_DIRECTORY}/game.html"

  GOOGLE_PLAY_PATH = 'https://play.google.com/store/apps/details?id=' # URL for Google Play
  
  # When scaling images, scale down to this width/height (whatever's smaller)
  # The image width (if a landscape image) is guaranteed to be 250px or less
  # The image height (if a portrait image) is guaranteed to be 250px or less
  MAX_SCREENSHOT_SIZE = 250 

  def build
    start = Time.new
    verify_files_exist
    load_data
    generate_site
    stop = Time.new
    puts "Done in #{stop - start} seconds."
  end

  private

  def load_data
    @games = JSON.parse(File.read(DATABASE_FILE))['games'].sort! {
      # sort by publication date, reverse chronologically
      |x, y| Time.parse(y['published']) <=> Time.parse(x['published'])
    }
    raise 'JSON structure changed; where is the top-level "games" list?' if @games.nil?
  end

  def verify_files_exist
    raise "#{DATABASE_FILE} not found" unless File.exist?(DATABASE_FILE)
    raise "#{TEMPLATE_DIRECTORY} directory not found" unless Dir.exist?(TEMPLATE_DIRECTORY)
  end

  def generate_site
    # Copy over CSS, fonts, JS, and the index page, plus site-wide images, etc.
    FileUtils.cp_r "#{TEMPLATE_DIRECTORY}/.", OUTPUT_DIR
    TEMPLATE_EXCLUSIONS.each do |exclusion|
      FileUtils.rm_rf exclusion
    end
    FileUtils.cp_r GAMES_DIR, OUTPUT_DIR

    generate_master_page
    generate_static_pages
    generate_front_page_game_entries
    generate_game_pages
  end

  def generate_game_pages
    @games.each do |g|
      html = File.read("#{GAME_PAGE_TEMPLATE}")
      html = html.gsub('@name', g['name']).gsub('@blurb', g['blurb'])

      # Start adding per-platform HTML
      # For HTML5 and Flash, add in-page game playing
      html = get_inpage_platforms_html(g, html)
      html = get_downloadable_platforms_html(g, html)
      html = get_mobile_links(g, html)
      html = get_screenshots(g, html)
      final_html = @master_page_html.sub(CONTENT_PLACEHOLDER, html).gsub('@title', g['name'])      
      filename = url_for_game(g)
      File.open("#{OUTPUT_DIR}/#{filename}", 'w') { |f| f.write(final_html) }
    end
  end

  # Replace @screenshots with screenshots
  def get_screenshots(g, html)    
    if !g['screenshots'].nil?
      ss_html = "<h2>Screenshots</h2>" # Mixing HTML and code is bad, dude.
      name = url_for_game(g).sub('.html', '')
      g['screenshots'].each do |s|
        url = "images/#{name}/#{s}"
        native_size = FastImage.size("data/#{url}") # [w, h]
        # Scale to 250px. Unless the image is already smaller. Then don't scale.
        scale = [1.0 * MAX_SCREENSHOT_SIZE / native_size[0],  1.0 * MAX_SCREENSHOT_SIZE / native_size[1], 1].min
        scale_w = (scale * native_size[0]).to_i
        scale_h = (scale * native_size[1]).to_i
        ss_html = "#{ss_html}<img src='#{url}' width='#{scale_w}' height='#{scale_h}' data-jslghtbx='#{url}' />"
      end
      html = html.gsub('@screenshots', ss_html)
    else
      html = html.gsub('@screenshots', '')
    end    
    return html
  end

  # Modifies "html": replaces @mobile with mobile links
  def get_mobile_links(g, html)    
    links_html = ''
    mobile_data = platform_data(g, ['android']) #TODO: iOS
    if !mobile_data.empty?      
      mobile_data.each do |platform, data|        
        link_target = "#{GOOGLE_PLAY_PATH}#{data}"
        links_html += "<a href='#{link_target}'><img src='images/google-play-badge.png' /></a>"
      end
      html = html.gsub('@mobile', links_html)
    else
      html = html.gsub('@mobile', '')
    end    
    return html
  end
  
  # Modifies "html": replaces @downloads with download links
  def get_downloadable_platforms_html(g, html)
    downloadable_data = platform_data(g, ['windows', 'linux', 'mac'])
    template = File.read("#{TEMPLATE_DIRECTORY}/snippets/download_game.html")
    downloads_html = ''

    if !downloadable_data.empty?
      downloadable_data.each do |platform, data|
        root_dir = GAMES_DIR.sub("#{DATA_DIR}/", '')
        url = "#{root_dir}/#{platform}/#{data}"
        name = "#{platform.capitalize} version"
        downloads_html = "#{downloads_html}#{template.gsub('@url', url).gsub('@name', name)}"
      end
    end

    # If we have something to download, show the download section.
    if !downloads_html.empty?
      downloads_section = File.read("#{TEMPLATE_DIRECTORY}/snippets/downloads.html")
      downloads_section.sub!('@html', downloads_html)
      html = html.sub('@downloads', downloads_section)
    else
      html = html.sub('@downloads', '')
    end

    return html
  end

  # Modifies "html": replaces @game with in-place game code (<object> for swf, <iframe> for HTML5)
  def get_inpage_platforms_html(g, html)
    in_page_data = platform_data(g, ['flash', 'html5'])

    if !in_page_data.empty?
      data = in_page_data[:html5] || in_page_data[:flash]  # html5 first, then flash -- not both
      platform = :html5 if in_page_data.key?(:html5)
      platform = :flash if platform.nil?

      # Common to flash/html5
      template = File.read("#{TEMPLATE_DIRECTORY}/snippets/#{platform}.html")
      template = template.gsub('@width', data['width'].to_s)
      template = template.gsub('@height', data['height'].to_s)

      if platform == :flash
        template = template.gsub('@swf', "games/flash/#{data['swf']}")
      elsif platform == :html5
        template = template.gsub('@folder', data['folder'])
      else
        raise "Not sure how to get in-page data for #{platform}"
      end
      html = html.sub('@game', template)
    else
      # get rid of @game if it's still around
      html = html.gsub('@game', '')
    end

    return html
  end

  def generate_front_page_game_entries
    featured_html = ''
    regular_html = ''
    platform_html = ''

    @games.each do |g|
      is_featured = g == @games[0] || g == @games[1]
      column_size = is_featured ? 6 : 4 # featured = half-screen, otherwise one-third
      # Regardless of extension, add size
      filename = g['screenshot']

      game_image = "#{IMAGES_DIR}/#{filename}"
      raise "Can't find image #{g['screenshot']} for #{g['name']} in #{IMAGES_DIR}" unless File.exist?(game_image)
      html = "<a href='#{url_for_game(g)}'><img class='img-responsive' src='#{game_image.sub('data/', '')}' /></a>"

      platform_html = ""
      g['platforms'].each do |platform_data|
        # TODO: switch to SVGs for these
        platform_data.each do |platform, data|
          # windows/linux: value = executable
          # android: value = google play ID
          # flash: value = { :width, :height, :swf }
          link_target = "#{GAMES_DIR}/#{platform}/#{data}" if ['windows', 'linux'].include?(platform)
          link_target = "#{GOOGLE_PLAY_PATH}#{data}" if platform == 'android'
          link_target = url_for_game(g) if ['flash', 'html5'].include?(platform)
          ext = platform == 'silverlight' ? 'png' : 'svg'
          platform_html = "#{platform_html}<a href='#{link_target}'><img src='images/#{platform}.#{ext}' width='32' height='32' /></a>"
        end
      end

      # Compose final HTML
      final_html = "<div class='col-sm-#{column_size}'>#{html}<br />#{platform_html}</div>"

      if (is_featured)
        featured_html = "#{featured_html}#{final_html}"
      else
        regular_html = "#{regular_html}#{final_html}"
      end
    end

    featured_html = File.read(JUMBOTRON_SNIPPET).sub('@content', featured_html)
    html = File.read("#{OUTPUT_DIR}/#{INDEX_PAGE}")
    html = @master_page_html.sub(CONTENT_PLACEHOLDER, "#{featured_html}#{regular_html}").gsub('@title', 'Home')
    File.write("#{OUTPUT_DIR}/#{INDEX_PAGE}", html)
    FileUtils.cp_r "#{IMAGES_DIR}/.", "#{OUTPUT_DIR}/images"
  end

  # Generates static pages from data/static_pages/*.md
  # Converts them into HTML, links them in the header
  def generate_static_pages
    raise 'Pages are not defined!' if @pages.nil?

    @pages.each do |p|
      markdown = File.read(p)
      to_html = Kramdown::Document.new(markdown).to_html
      html = @master_page_html.sub(CONTENT_PLACEHOLDER, to_html).gsub('@title', to_title(get_page_name(p)))
      page_name = get_page_name(p)
      File.write("#{OUTPUT_DIR}/#{page_name}.html", html)
    end

    puts "Generated #{@pages.count} pages."
  end

  # Generates the "master page" which contains all the common information
  # (eg. header, footer, navbar); content plugs into this. For now, we're
  # using/abusing this by calling it index.html
  def generate_master_page
    # Get a list of all static pages. We need links for our header.
    index_page = File.read("#{OUTPUT_DIR}/#{INDEX_PAGE}")
    # Naively, sort alphabetically. That usually makes sense.
    @pages = Dir.glob("#{STATIC_PAGES_DIR}/*.md").sort
    navbar_template = File.read(NAVBAR_LINK_SNIPPET)
    links_html = ''

    @pages.each do |page|
      # Create the header link for this page
      # Creates relative links. This is okay, since our site is flat (no subdirectories)
      page_name = get_page_name(page)
      html = navbar_template.gsub('@url', "#{page_name}.html").gsub('@title', to_title(page_name))
      links_html = "#{links_html}#{html}"
    end

    @master_page_html = index_page.gsub(NAVBAR_LINKS_PLACEHOLDER, links_html)
  end

### start: make a Game class and put these inside

   # page file: eg. data/pages/privacy_policy.md
   # returns: 'privacy_policy'
  def get_page_name(markdown_filename)
    name_start = markdown_filename.rindex('/') + 1
    name_stop = markdown_filename.rindex('.md')
    page_name = markdown_filename[name_start, name_stop - name_start]
    return page_name
  end

  # g => { :name => 'Quest for the Royal Jelly'}
  # return: quest-for-the-royal-jelly.html
  def url_for_game(g)
    name = g['name']
    name = name.gsub(' ', '-').gsub('_', '-').downcase.strip.chomp
    return "#{name}.html"
  end

  # Get all datapoints for specific platforms. If you pass in (g, ['windows', 'linux']),
  # you'll get data for both windows and linux (if they're both there).
  # HTML5 and Flash need to be shown in-page.
  # Windows, Linux, and Mac need a download link.
  # Returns an array, eg. {:windows => ..., :linux => ...}
  def platform_data(g, target_platforms)
    to_return = {}

    g['platforms'].each do |platform_data|
      platform_data.each do |platform, data|
        # if you specify both, returns the first one found
        to_return[platform.to_sym] = data if target_platforms.include?(platform)
      end
    end

    return to_return
  end

### end: make a Game class and put these inside

  # privacy_policy => Privacy Policy
  # who_is_that_person => Who is that Person
  def to_title(sentence)
    stop_words = %w{a an and the or for of nor} #there is no such thing as a definite list of stop words, so you may edit it according to your needs.
    sentence.gsub('_', ' ').split.each_with_index.map{|word, index| stop_words.include?(word) && index > 0 ? word : word.capitalize }.join(" ")
  end
end

File: src/builder.py
import distutils.dir_util
import glob
import io
import json
import os
import shutil
import time

# Custom code
import src.file_io as file_io

# Extensions
import markdown
from PIL import Image

class Builder:

    OUTPUT_DIR = 'bin' # where to build to
    DATA_DIR = 'data' # where our data files are
    DATABASE_FILE = os.path.join(DATA_DIR, "games.json")
    STATIC_PAGES_DIR = os.path.join(DATA_DIR, "static_pages")
    IMAGES_DIR = os.path.join(DATA_DIR, "images")
    GAMES_DIR = "games"
    GUIDES_DIR = "guides"

    TEMPLATE_DIRECTORY = 'templates'
    # We copy the template dir. But not these items.
    TEMPLATE_EXCLUSIONS = ['snippets']
    # Copy these from the root to bin
    INDIVIDUAL_FILES_TO_COPY = ['CNAME', 'favicon.ico']

    INDEX_PAGE = 'index.html' # the home index page; this is also the layout.
    NAVBAR_LINKS_PLACEHOLDER = '<!-- DG navbar links -->' # Where the navbar list of pages goes
    # Snippets where we put in real data (eg. game page URLs)
    NAVBAR_LINK_SNIPPET = os.path.join(TEMPLATE_DIRECTORY, "snippets", "navbar_link.html")
    JUMBOTRON_SNIPPET = os.path.join(TEMPLATE_DIRECTORY, "snippets", "jumbotron.html")
    SCREENSHOTS_SNIPPET = os.path.join(TEMPLATE_DIRECTORY, "snippets", "screenshots.html")
    CONTENT_PLACEHOLDER = '<!-- DG content -->' # Where in the template we fill in the page content
    GAME_PAGE_TEMPLATE = os.path.join(TEMPLATE_DIRECTORY, "game.html")

    GOOGLE_PLAY_PATH = 'https://play.google.com/store/apps/details?id=' # URL for Google Play

    ITCH_IO_URL_ROOT = 'https://deengames.itch.io'
    STEAM_ROOT_URL = 'https://store.steampowered.com/app'

    # When scaling images, scale down to this width/height (whatever's smaller)
    # The image width (if a landscape image) is guaranteed to be 250px or less
    # The image height (if a portrait image) is guaranteed to be 250px or less
    MAX_SCREENSHOT_SIZE = 250

    def build(self):
        start = time.time()
        self.__verify_files_exist()
        self.__load_data()
        self.__generate_site()
        stop = time.time()
        print("Done in {0} seconds.".format(stop - start))

    def __load_data(self):
        raw_json = file_io.read(Builder.DATABASE_FILE)
        json_games = json.loads(raw_json)['games']
        
        if json_games == None:
            raise(Exception('JSON structure changed; where is the top-level "games" list?'))

        # sort by publication date, reverse chronologically (newest first)
        json_games.sort(key = lambda x: x["published"], reverse = True)

        # Convert into Game instances
        self.games = []
        for j in json_games:
            if "ignore" in j and j["ignore"] != "false":
                print("Skipping {}".format(j["name"]))
            else:
                self.games.append(Game(j)) 

    def __verify_files_exist(self):
        if not os.path.isfile(Builder.DATABASE_FILE):
            raise(Exception("{0} not found".format(Builder.DATABASE_FILE)))
        if not os.path.isdir(Builder.TEMPLATE_DIRECTORY):
            raise(Exception("{0} directory not found".format(Builder.TEMPLATE_DIRECTORY)))

    def __generate_site(self):
        # Delete the old output (clean build)
        shutil.rmtree(Builder.OUTPUT_DIR, True)
        
        # Copy over CSS, fonts, JS, and the index page, plus site-wide images, etc.
        distutils.dir_util.copy_tree(Builder.TEMPLATE_DIRECTORY, Builder.OUTPUT_DIR)
        for exclusion in Builder.TEMPLATE_EXCLUSIONS:                
            shutil.rmtree(os.path.join(Builder.OUTPUT_DIR, exclusion), True)

        # Copy individual files
        for item in Builder.INDIVIDUAL_FILES_TO_COPY:
            shutil.copyfile(item, os.path.join(Builder.OUTPUT_DIR, item))

        self.__generate_master_page()
        self.__generate_static_pages()
        self.__generate_front_page_game_entries()

    def __generate_front_page_game_entries(self):
        featured_html = ''
        regular_html = ''
        platform_html = ''
        in_this_row = 0

        for g in self.games:
            is_featured = g == self.games[0]
             
            # featured = full-screen, otherwise one-third
            # also, featured image requires center-block class to center. Otherwise, whitespace on the RHS.
            column_size = 12 if is_featured else 4
            image_class = 'img-responsive center-block' if is_featured else 'img-responsive'
            # Regardless of extension, add size
            filename = g.get('screenshot')

            game_image = os.path.join(Builder.IMAGES_DIR, filename)
            if not os.path.isfile(game_image):
                raise(Exception("Can't find image {2}/{0} for game {1}".format(g.get('screenshot'), g.get('name'), Builder.IMAGES_DIR)))
            html = "<a href='{0}'><img class='{1}' src='{2}' /></a>".format(g.get_url(), image_class, game_image.replace('{0}{1}'.format(Builder.DATA_DIR, os.sep), ''))

            platform_html = ""
            # Some games have only a custom URL and no platforms; ignore those.
            if g.has("platforms"):
                for platform_data in g.get('platforms'):
                    for platform, data in platform_data.items():
                        # windows/linux: value = executable
                        # android: value = google play ID
                        # flash: value = { :width, :height, :swf }
                        # silverlight: value = xap file
                        if 'windows' in platform or 'linux' in platform or 'mac' in platform:
                            link_target = "{0}/{1}/{2}".format(Builder.GAMES_DIR, platform, data)
                        elif 'android' in platform:
                            link_target = "{0}{1}".format(Builder.GOOGLE_PLAY_PATH, data)
                        elif 'flash' in platform or 'html5' in platform or 'silverlight' in platform:
                            link_target = g.get_url()
                        else:
                            print("WARNING: Not sure what the link target is for {0}".format(platform))
                        
                        ext = ('png' if platform == 'silverlight' else 'svg')
                        platform_html = "{0}<img src='images/{1}.{2}' width='32' height='32' />".format(platform_html, platform, ext)

            # Compose final HTML
            final_html = "<div class='col-sm-{0}'>{1}{2}</div>".format(column_size, html, platform_html)

            if not is_featured:
                if in_this_row == 0:
                    final_html = "{0}{1}".format("<div class='row'>", final_html)
                elif in_this_row == 2:
                    final_html = "{0}{1}".format(final_html, "</div>")

                in_this_row += 1
                in_this_row %= 3

            if is_featured:
                featured_html = "{0}\r\n{1}".format(featured_html, final_html)
            else:
                regular_html = "{0}\r\n{1}".format(regular_html, final_html)

            

        # close unopened div if number of games is not divisible by 3
        if (in_this_row > 0):
            regular_html += "</div>"

        featured_html = file_io.read(Builder.JUMBOTRON_SNIPPET).replace('@content', featured_html)
        html = file_io.read(os.path.join(Builder.OUTPUT_DIR, Builder.INDEX_PAGE))
        html = self.master_page_html.replace(Builder.CONTENT_PLACEHOLDER, "{0}{1}".format(featured_html, regular_html)).replace('@title', 'Home')
        file_io.write(os.path.join(Builder.OUTPUT_DIR, Builder.INDEX_PAGE), html)
        distutils.dir_util.copy_tree(os.path.join(Builder.IMAGES_DIR, "."), os.path.join(Builder.OUTPUT_DIR, "images"))

    # Generates static pages from data/static_pages/*.md
    # Converts them into HTML, links them in the header
    def __generate_static_pages(self):
        if self.pages == None:
                raise(Exception('Pages are not defined!'))

        for p in self.pages:
            raw_markdown = file_io.read(p)
            to_html = markdown.markdown(raw_markdown)
            html = self.master_page_html.replace(Builder.CONTENT_PLACEHOLDER, to_html).replace('@title', self.__to_title(self.__get_page_name(p)))
            page_name = self.__get_page_name(p)

            file_io.write(os.path.join(Builder.OUTPUT_DIR, "{0}.html".format(page_name)), html)

        print("Generated {0} games and {1} static pages.".format(len(self.games), len(self.pages)))

    # Generates the "master page" which contains all the common information
    # (eg. header, footer, navbar); content plugs into this. For now, we're
    # using/abusing this by calling it index.html
    def __generate_master_page(self):
        # Get a list of all static pages. We need links for our header.
        index_page = file_io.read(os.path.join(Builder.OUTPUT_DIR, Builder.INDEX_PAGE))
        # Naively, sort alphabetically. That usually makes sense.
        self.pages = glob.glob(os.path.join(Builder.STATIC_PAGES_DIR, "*.md"))
        self.pages.sort()

        navbar_template = file_io.read(Builder.NAVBAR_LINK_SNIPPET)
        
        links_html = ''

        for page in self.pages:
            if "\\draft" in page:
                print("Skipping static page draft for {}".format(page))
            else:
                # Create the header link for this page
                # Creates relative links. This is okay, since our site is flat (no subdirectories)
                page_name = self.__get_page_name(page)
                html = navbar_template.replace('@url', "{0}.html".format(page_name)).replace('@title', self.__to_title(page_name))
                links_html = "{0}{1}".format(links_html, html)

        self.master_page_html = index_page.replace(Builder.NAVBAR_LINKS_PLACEHOLDER, links_html)

     # page file: eg. data/pages/privacy_policy.md
     # returns: 'privacy_policy'
    def __get_page_name(self, markdown_filename):
        name_start = markdown_filename.rindex(os.sep) + 1
        name_stop = markdown_filename.rindex('.md')
        page_name = markdown_filename[name_start:name_stop]
        return page_name

    # privacy_policy => Privacy Policy
    # who_is_that_person => Who is that Person
    def __to_title(self, sentence):
        stop_words = ['a', 'an', 'and', 'the', 'or', 'for', 'of', 'nor'] #there is no such thing as a definite list of stop words, so you may edit it according to your needs.
        words = sentence.replace('_', ' ').split()

        title = ""
        for word in words:
            if word not in stop_words:
                word = word.capitalize()
            title = "{0}{1} ".format(title, word)
        return title.strip()

class Game:
    def __init__(self, json):
        self.json = json

    def has(self, key):
        return key in self.json

    def get(self, key):
        return self.json[key]
    
    # Get all datapoints for specific platforms. If you pass in (['windows', 'linux']),
    # you'll get data for both windows and linux (if the game has data for both).
    # HTML5, Flash, and Silverlight need to be shown in-page.
    # Windows, Linux, and Mac need a download link.
    # Returns an array, eg. {:windows => ..., :linux => ...}
    def platform_data(self, target_platforms):
        to_return = {}

        for platform_data in self.get('platforms'):
            for platform, data in platform_data.items():
                # if you specify multiple, returns the first one found
                if platform in target_platforms:
                    to_return[platform] = data

        return to_return   

    ### REGION: URL and HTML responsibilities ###

    
    # HTML for @game for in-place games (Flash, HTML5/JS, Silverlight).
    # Generates <object> for Flash/Silverlight, <iframe> for HTML5
    def get_inpage_platforms_html(self, template_directory):
        in_page_data = self.platform_data(['flash', 'html5', 'silverlight'])

        if in_page_data: # not empty
            platform = None
            # html5 first, then flash -- not both
            if "html5" in in_page_data:
                platform = "html5" 
                data = in_page_data["html5"]
            if "flash" in in_page_data:
                platform = "flash"
                data = in_page_data["flash"]
            if "silverlight" in in_page_data:
                platform = "silverlight" 
                data = in_page_data["silverlight"]
            if platform == None:
                raise(Exception("Not sure how to process platform: {0}".format(in_page_data)))
            
            # Common to flash/html5/silverlight
            template = file_io.read(os.path.join(template_directory, "snippets", "{0}.html".format(platform)))
            template = template.replace('@width', str(data['width']))
            template = template.replace('@height', str(data['height']))

            if platform == "flash":
                template = template.replace('@swf', "games/flash/{0}".format(data['swf']))
            elif platform == "html5":
                template = template.replace('@folder', data['folder'])
            elif platform == "silverlight":
                template = template.replace('@xap', "games/silverlight/{0}".format(data['xap']))
            else:
                raise(Exception("Not sure how to get in-page data for {0}".format(platform)))
            
            return template
        else:
            return "";        


    # Download links for @downloads, if applicable (desktop only)
    def get_downloadable_platforms_html(self, template_directory, games_directory):
        downloadable_data = self.platform_data(['windows', 'linux', 'mac'])
        template = file_io.read(os.path.join(template_directory, "snippets", "download_game.html"))
        downloads_html = ''

        if downloadable_data: # not empty
            for platform, data in downloadable_data.items():
                url = "{0}/{1}/{2}".format(games_directory, platform, data)
                name = "{0} version".format(platform.capitalize())
                downloads_html = "{0}{1}".format(downloads_html, template.replace('@url', url).replace('@name', name))

        # If we have something to download, show the download section. (Separate template.)
        if downloads_html: # not empty
            downloads_section = file_io.read(os.path.join(template_directory, "snippets", "downloads.html"))
            downloads_section = downloads_section.replace('@html', downloads_html)
            return downloads_section
        else:
            return ""


    # Mobile links for @mobile, if applicable (mobile apps only)
    def get_mobile_links(self, google_play_prefix):
        links_html = ''
        mobile_data = self.platform_data(['android']) #TODO: iOS
        if mobile_data: # not empty
            for platform, data in mobile_data.items():
                link_target = "{0}{1}".format(google_play_prefix, data)
                links_html += "<a href='{0}'><img src='images/google-play-badge.png' /></a>".format(link_target)
            
            return links_html
        else:
            return ""

    # Screenshot HTML for @screenshots
    def get_screenshots(self):
        if self.has('screenshots'):
            template = file_io.read(Builder.SCREENSHOTS_SNIPPET)
            name = self.get_url().replace('.html', '')
            ss_html = ''
            for s in self.get('screenshots'):
                url = os.path.join("images", name, s)
                native_size = Image.open(os.path.join("data", url)).size # (w, h)
                # Scale to 250px. Unless the image is already smaller. Then don't scale.
                scale = min(1.0 * Builder.MAX_SCREENSHOT_SIZE / native_size[0], 1.0 * Builder.MAX_SCREENSHOT_SIZE / native_size[1], 1)
                scale_w = int(scale * native_size[0])
                scale_h = int(scale * native_size[1])
                ss_html = "{0}<img src='{1}' width='{2}' height='{3}' data-jslghtbx='{1}' class='screenshot' />".format(ss_html, url, scale_w, scale_h)
            
            return template.replace('@screenshots', ss_html)
        else:
            return ""        

    # return: Quest for the Royal Jelly => quest-for-the-royal-jelly
    def get_url(self):
        if self.has("steamAppId"):
            steamAppId = self.get("steamAppId")
            return "{}/{}".format(Builder.STEAM_ROOT_URL, steamAppId)
        elif self.has("customUrl"):
            return self.get("customUrl")
        else:
            name = self.get('name')
            name = name.replace(' ', '-').replace('_', '-').replace("'", "").lower().strip()
            return "{}/{}".format(Builder.ITCH_IO_URL_ROOT, name)

    def get_educators_guide_html(self):
        if self.has('educators_guide'):
            return "<br /><a href='guides/{0}'><img src='images/educators_guide.svg' width='32' height='32' /> Parents/Educators Guide</a>".format(self.get('educators_guide'))
        else:
            return ""
    